File: android/app/src/main/kotlin/ai/gowtam/oak/ui/TypeBadge.kt
package ai.gowtam.oak.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp

/**
 * A labeled Pokémon type chip — color **and** text, never color-only.
 *
 * Enamel & Paper: tinted pill (16/72/30 mix via [OakType.badgeFill] /
 * [OakType.badgeInk] / [OakType.badgeBorder]), not a Signal solid. Nunito Sans
 * 600 / 11sp, pill radius. The type name is always shown, so color is never the
 * sole carrier of meaning; the label scales with the user's font-size (`sp`).
 */
@Composable
fun TypeBadge(
    type: String,
    modifier: Modifier = Modifier,
) {
    val oak = LocalOakColors.current
    val dark = isSystemInDarkTheme()
    val surface = MaterialTheme.colorScheme.surface
    val fill = OakType.badgeFill(type, surface, dark)
    val ink = OakType.badgeInk(type, oak.textStrong, dark)
    val stroke = OakType.badgeBorder(type)
    val label = type.trim().replaceFirstChar { it.uppercase() }
    val chipShape = RoundedCornerShape(OakRadius.pill)

    Text(
        text = label,
        color = ink,
        style = MaterialTheme.typography.labelSmall.copy(
            fontFamily = NunitoSansFamily,
            fontWeight = FontWeight.SemiBold,
            fontSize = 11.sp,
            letterSpacing = 0.03.em,
        ),
        textAlign = TextAlign.Center,
        maxLines = 1,
        modifier = modifier
            .background(fill, chipShape)
            .border(width = 1.dp, color = stroke, shape = chipShape)
            .padding(horizontal = 10.dp, vertical = 2.dp)
            .semantics { contentDescription = "$label type" },
    )
}

File: ios/OakApp/UI/TypeBadge.swift
import SwiftUI

/// A labeled Pokémon type chip — color **and** text, never color-only.
///
/// Enamel & Paper recipe: a **tinted pill** from ``Theme.TypeBadgeChrome``
/// (type mixed into surface / textStrong, 30% type border) — not the Signal
/// solid chip. The palette is sourced from `Theme.type(_:)` (the 18 type
/// solids), so the same slug renders consistently everywhere it appears.
///
/// The type's name is always shown as text, so color is not the sole carrier of
/// meaning (M-AC-UI9.3). Typography is Nunito Sans 600 / 11 (`Theme.body`
/// caption2 semibold) — a Dynamic Type text style, so the chip grows with the
/// user's preferred size instead of clipping (M-AC-UI9.2).
struct TypeBadge: View {
  /// The lowercase type slug, e.g. `"fire"` (one of the 18 `TYPE_NAMES`).
  let type: String

  /// The user-facing label, capitalized from the slug (`"fire"` → `"Fire"`).
  private var label: String {
    type.capitalized
  }

  var body: some View {
    Text(label)
      .font(Theme.body(.caption2, weight: .semibold))
      .tracking(0.33)
      .lineLimit(1)
      .padding(.horizontal, 10)
      .padding(.vertical, 2)
      .foregroundStyle(Theme.TypeBadgeChrome.ink(type))
      .background(Theme.TypeBadgeChrome.fill(type), in: Capsule())
      .overlay {
        Capsule().strokeBorder(Theme.TypeBadgeChrome.border(type), lineWidth: 1)
      }
      .accessibilityElement(children: .ignore)
      .accessibilityLabel("\(label) type")
  }
}

#Preview("Type badges") {
  let types = [
    "normal", "fire", "water", "electric", "grass", "ice",
    "fighting", "poison", "ground", "flying", "psychic", "bug",
    "rock", "ghost", "dragon", "dark", "steel", "fairy",
  ]
  return ScrollView {
    LazyVGrid(columns: Array(repeating: GridItem(.flexible()), count: 3), spacing: 12) {
      ForEach(types, id: \.self) { TypeBadge(type: $0) }
    }
    .padding()
  }
}
